gnrc_ipv6_simple_subnets: auto-configuration for nested subnets on a simple tree topology

benpicco opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

 This module provides an automatic configuration for networks with a simple  tree topology.
 
 If a sufficiently large IPv6 prefix is provided via Router Advertisements, a  routing node with this module will automatically configure subnets from it for each downstream interface.
 
 There can only be a single routing node on each level of the network but an arbitrary number of leaf nodes.
 
![](https://raw.githubusercontent.com/RIOT-OS/RIOT/98db19c53d5c04937b4ae03ec61c749805f474ab/doc/doxygen/src/gnrc_ipv6_auto_subnets_simple.svg)
 
 The downstream network(s) get the reduced prefix via Router Advertisements and the process repeats until the bits of prefix are exhausted. (The smallest network must still have a /64 prefix.)
 
 The downstream router will send an otherwise empty router advertisement with only the Route Information Option set to the upstream network.
 The Route Information Option contains the prefix of the downstream network so that upstream hosts will no longer consider hosts in this subnet on-link but instead will use the downstream router to route to the new subnet.

### Testing procedure

I daisy-chained three boards (one `same54-xpro` and two `weact-f411ce`).
The `same54-xpro` uses Ethernet as the uplink interface and SLIP for downlink (using #16530 and #16531).
The WeAct boards have two SLIP interfaces each:

```C
#define SLIPDEV_PARAMS          { .uart = UART_DEV(0),   \
                                  .baudrate = SLIPDEV_PARAM_BAUDRATE }, \
                                { .uart = UART_DEV(1),   \
                                  .baudrate = SLIPDEV_PARAM_BAUDRATE }
```

This was then tested with `examples/gnrc_networking` and 

```make
USEMODULE += slipdev
USEMODULE += gnrc_ipv6_auto_subnets
```

The `same54-xpro` was configured with

```
USEMODULE += slipdev
USEMODULE += gnrc_dhcpv6_client_6lbr
CFLAGS += -DSLIPDEV_PARAM_UART=UART_DEV\(1\)
```

<details><summary>first level node (gateway)</summary>

```
2021-06-07 18:23:51,539 # Iface  6 
2021-06-07 18:23:51,543 #           Long HWaddr: E6:EA:AF:F8:AF:5D:EB:E5 
2021-06-07 18:23:51,547 #           MTU:65535  HL:64  RTR  
2021-06-07 18:23:51,548 #           RTR_ADV  
2021-06-07 18:23:51,551 #           Link type: wired
2021-06-07 18:23:51,556 #           inet6 addr: fe80::e4ea:aff8:af5d:ebe5  scope: link  VAL
2021-06-07 18:23:51,564 #           inet6 addr: 2001:16b8:4569:88fc:e4ea:aff8:af5d:ebe5  scope: global  VAL
2021-06-07 18:23:51,566 #           inet6 group: ff02::2
2021-06-07 18:23:51,569 #           inet6 group: ff02::1
2021-06-07 18:23:51,573 #           inet6 group: ff02::1:ff5d:ebe5
2021-06-07 18:23:51,575 #           inet6 group: ff02::1a
2021-06-07 18:23:51,577 #           
2021-06-07 18:23:51,579 #           Statistics for Layer 2
2021-06-07 18:23:51,582 #             RX packets 2  bytes 168
2021-06-07 18:23:51,587 #             TX packets 11 (Multicast: 0)  bytes 1038
2021-06-07 18:23:51,590 #             TX succeeded 0 errors 0
2021-06-07 18:23:51,593 #           Statistics for IPv6
2021-06-07 18:23:51,596 #             RX packets 2  bytes 168
2021-06-07 18:23:51,601 #             TX packets 11 (Multicast: 11)  bytes 1038
2021-06-07 18:23:51,604 #             TX succeeded 11 errors 0
2021-06-07 18:23:51,604 # 
2021-06-07 18:23:51,607 # Iface  5  HWaddr: FC:C2:3D:0D:2D:1F 
2021-06-07 18:23:51,612 #           L2-PDU:1500  MTU:1492  HL:255  RTR  
2021-06-07 18:23:51,615 #           Source address length: 6
2021-06-07 18:23:51,617 #           Link type: wired
2021-06-07 18:23:51,623 #           inet6 addr: fe80::fec2:3dff:fe0d:2d1f  scope: link  VAL
2021-06-07 18:23:51,630 #           inet6 addr: 2001:16b8:4569:8800:fec2:3dff:fe0d:2d1f  scope: global  VAL
2021-06-07 18:23:51,633 #           inet6 group: ff02::2
2021-06-07 18:23:51,635 #           inet6 group: ff02::1
2021-06-07 18:23:51,639 #           inet6 group: ff02::1:ff0d:2d1f
2021-06-07 18:23:51,640 #           
2021-06-07 18:23:51,643 #           Statistics for Layer 2
2021-06-07 18:23:51,646 #             RX packets 55  bytes 12315
2021-06-07 18:23:51,651 #             TX packets 8 (Multicast: 8)  bytes 734
2021-06-07 18:23:51,654 #             TX succeeded 8 errors 0
2021-06-07 18:23:51,657 #           Statistics for IPv6
2021-06-07 18:23:51,660 #             RX packets 16  bytes 4722
2021-06-07 18:23:51,664 #             TX packets 8 (Multicast: 8)  bytes 622
2021-06-07 18:23:51,667 #             TX succeeded 8 errors 0
2021-06-07 18:23:51,668 # 
> nib prefix
2021-06-07 18:23:59,163 # nib prefix
2021-06-07 18:23:59,169 # 2001:16b8:4569:88fc::/62 dev #6  expires 7187 sec deprecates 3587 sec
2021-06-07 18:23:59,175 # 2001:16b8:4569:8800::/64 dev #5  expires 7188 sec deprecates 3588 sec
```
</details>

<details><summary>second level node</summary>

```
2021-06-07 18:23:53,306 # Iface  7 
2021-06-07 18:23:53,306 #           Long HWaddr: 6E:A6:68:BD:60:63:21:D5 
2021-06-07 18:23:53,307 #           MTU:65535  HL:64  RTR  
2021-06-07 18:23:53,307 #           RTR_ADV  
2021-06-07 18:23:53,307 #           Link type: wired
2021-06-07 18:23:53,308 #           inet6 addr: fe80::6ca6:68bd:6063:21d5  scope: link  VAL
2021-06-07 18:23:53,309 #           inet6 addr: 2001:16b8:4569:88fc:6ca6:68bd:6063:21d5  scope: global  VAL
2021-06-07 18:23:53,310 #           inet6 group: ff02::2
2021-06-07 18:23:53,310 #           inet6 group: ff02::1
2021-06-07 18:23:53,311 #           inet6 group: ff02::1:ff63:21d5
2021-06-07 18:23:53,311 #           
2021-06-07 18:23:53,311 #           Statistics for Layer 2
2021-06-07 18:23:53,312 #             RX packets 12  bytes 1138
2021-06-07 18:23:53,312 #             TX packets 3 (Multicast: 0)  bytes 232
2021-06-07 18:23:53,313 #             TX succeeded 0 errors 0
2021-06-07 18:23:53,313 #           Statistics for IPv6
2021-06-07 18:23:53,314 #             RX packets 12  bytes 1138
2021-06-07 18:23:53,314 #             TX packets 3 (Multicast: 3)  bytes 232
2021-06-07 18:23:53,315 #             TX succeeded 3 errors 0
2021-06-07 18:23:53,315 # 
2021-06-07 18:23:53,315 # Iface  6 
2021-06-07 18:23:53,316 #           Long HWaddr: 6E:A6:68:BD:60:63:21:D4 
2021-06-07 18:23:53,316 #           MTU:65535  HL:64  RTR  
2021-06-07 18:23:53,316 #           RTR_ADV  
2021-06-07 18:23:53,317 #           Link type: wired
2021-06-07 18:23:53,318 #           inet6 addr: fe80::6ca6:68bd:6063:21d4  scope: link  VAL
2021-06-07 18:23:53,319 #           inet6 addr: 2001:16b8:4569:88fe:6ca6:68bd:6063:21d4  scope: global  VAL
2021-06-07 18:23:53,319 #           inet6 group: ff02::2
2021-06-07 18:23:53,320 #           inet6 group: ff02::1
2021-06-07 18:23:53,320 #           inet6 group: ff02::1:ff63:21d4
2021-06-07 18:23:53,320 #           
2021-06-07 18:23:53,321 #           Statistics for Layer 2
2021-06-07 18:23:53,321 #             RX packets 1  bytes 64
2021-06-07 18:23:53,322 #             TX packets 3 (Multicast: 0)  bytes 232
2021-06-07 18:23:53,322 #             TX succeeded 0 errors 0
2021-06-07 18:23:53,323 #           Statistics for IPv6
2021-06-07 18:23:53,323 #             RX packets 1  bytes 64
2021-06-07 18:23:53,324 #             TX packets 3 (Multicast: 3)  bytes 232
2021-06-07 18:23:53,325 #             TX succeeded 3 errors 0
2021-06-07 18:23:53,325 # 
> nib prefix
2021-06-07 18:23:56,314 # nib prefix
2021-06-07 18:23:56,316 # 2001:16b8:4569:88fc::/62 dev #7  expires 7190 sec deprecates 3590 sec
2021-06-07 18:23:56,317 # 2001:16b8:4569:88fe::/63 dev #6  expires 7190 sec deprecates 3590 sec
```
</details>

<details><summary>third level node</summary>

```
2021-06-07 18:24:01,378 # Iface  7 
2021-06-07 18:24:01,379 #           Long HWaddr: BE:4E:31:E9:94:58:5F:37 
2021-06-07 18:24:01,380 #           MTU:65535  HL:64  RTR  
2021-06-07 18:24:01,380 #           RTR_ADV  
2021-06-07 18:24:01,381 #           Link type: wired
2021-06-07 18:24:01,381 #           inet6 addr: fe80::bc4e:31e9:9458:5f37  scope: link  VAL
2021-06-07 18:24:01,382 #           inet6 addr: 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37  scope: global  VAL
2021-06-07 18:24:01,383 #           inet6 group: ff02::2
2021-06-07 18:24:01,383 #           inet6 group: ff02::1
2021-06-07 18:24:01,384 #           inet6 group: ff02::1:ff58:5f37
2021-06-07 18:24:01,384 #           
2021-06-07 18:24:01,384 #           Statistics for Layer 2
2021-06-07 18:24:01,385 #             RX packets 0  bytes 0
2021-06-07 18:24:01,385 #             TX packets 5 (Multicast: 0)  bytes 344
2021-06-07 18:24:01,386 #             TX succeeded 0 errors 0
2021-06-07 18:24:01,386 #           Statistics for IPv6
2021-06-07 18:24:01,386 #             RX packets 0  bytes 0
2021-06-07 18:24:01,387 #             TX packets 5 (Multicast: 5)  bytes 344
2021-06-07 18:24:01,387 #             TX succeeded 5 errors 0
2021-06-07 18:24:01,388 # 
2021-06-07 18:24:01,388 # Iface  6 
2021-06-07 18:24:01,389 #           Long HWaddr: BE:4E:31:E9:94:58:5F:36 
2021-06-07 18:24:01,389 #           MTU:65535  HL:64  RTR  
2021-06-07 18:24:01,390 #           RTR_ADV  
2021-06-07 18:24:01,391 #           Link type: wired
2021-06-07 18:24:01,392 #           inet6 addr: fe80::bc4e:31e9:9458:5f36  scope: link  VAL
2021-06-07 18:24:01,393 #           inet6 addr: 2001:16b8:4569:88fe:bc4e:31e9:9458:5f36  scope: global  VAL
2021-06-07 18:24:01,394 #           inet6 group: ff02::2
2021-06-07 18:24:01,394 #           inet6 group: ff02::1
2021-06-07 18:24:01,395 #           inet6 group: ff02::1:ff58:5f36
2021-06-07 18:24:01,395 #           
2021-06-07 18:24:01,395 #           Statistics for Layer 2
2021-06-07 18:24:01,396 #             RX packets 9  bytes 604
2021-06-07 18:24:01,397 #             TX packets 8 (Multicast: 0)  bytes 500
2021-06-07 18:24:01,397 #             TX succeeded 0 errors 0
2021-06-07 18:24:01,398 #           Statistics for IPv6
2021-06-07 18:24:01,398 #             RX packets 9  bytes 604
2021-06-07 18:24:01,399 #             TX packets 8 (Multicast: 5)  bytes 500
2021-06-07 18:24:01,399 #             TX succeeded 8 errors 0
2021-06-07 18:24:01,399 # 
> nib prefix
2021-06-07 18:24:04,386 # nib prefix
2021-06-07 18:24:04,388 # 2001:16b8:4569:88fe::/63 dev #6  expires 7181 sec deprecates 3581 sec
2021-06-07 18:24:04,390 # 2001:16b8:4569:88ff::/64 dev #7  expires 7181 sec deprecates 3581 sec
> ping 2600::
2021-06-07 18:24:07,754 # ping 2600::
2021-06-07 18:24:07,896 # 12 bytes from 2600::: icmp_seq=0 ttl=51 time=141.650 ms
2021-06-07 18:24:08,897 # 12 bytes from 2600::: icmp_seq=1 ttl=51 time=141.296 ms
2021-06-07 18:24:09,896 # 12 bytes from 2600::: icmp_seq=2 ttl=51 time=141.132 ms
2021-06-07 18:24:09,896 # 
2021-06-07 18:24:09,896 # --- 2600:: PING statistics ---
2021-06-07 18:24:09,897 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-06-07 18:24:09,898 # round-trip min/avg/max = 141.132/141.359/141.650 ms
```
</details>

The deepest node is reachable from an outside host

```
% ping 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37
PING 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37(2001:16b8:4569:88ff:bc4e:31e9:9458:5f37) 56 data bytes
64 bytes from 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37: icmp_seq=1 ttl=62 time=41.4 ms
64 bytes from 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37: icmp_seq=2 ttl=62 time=40.2 ms
64 bytes from 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37: icmp_seq=3 ttl=62 time=40.2 ms
^C
--- 2001:16b8:4569:88ff:bc4e:31e9:9458:5f37 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 40.226/40.632/41.427/0.561 ms
```

### Issues/PRs references

depends on ~~#16568~~, ~~#16672~~, ~~#16557~~, ~~#16671~~

miri64 commented on sys/net/gnrc/network_layer/ipv6/nib/nib.c:
```diff
@@ -1490,6 +1490,62 @@ static inline bool _multihop_p6c(gnrc_netif_t *netif, _nib_abr_entry_t *abr)
+    while ((downstream = gnrc_netif_iter(downstream))) {
+        ipv6_addr_t new_pfx;
+
+        if (downstream == upstream) {
+            continue;
+        }
```
What if there are supposed to be multiple `upstreams`?

benpicco replied:
That's not supported - but it's also not supported by `gnrc_dhcpv6_client_6lbr`.

miri64 replied:
Shouldn't it be possible when using multiple clients, one for each interface?

miri64 commented on makefiles/pseudomodules.inc.mk:
```diff
@@ -41,6 +41,7 @@ PSEUDOMODULES += gnrc_ipv6_nib_6ln
 PSEUDOMODULES += gnrc_ipv6_nib_6lr
 PSEUDOMODULES += gnrc_ipv6_nib_dns
 PSEUDOMODULES += gnrc_ipv6_nib_router
+PSEUDOMODULES += gnrc_ipv6_nib_subnets
```
I'd prefer this its own "routing protocol" rather than a sub-module of the NIB. Making it a sub-module makes it look like it is an "official" feature of IPv6, while this more of a flashy hack than anything else. It should be doable with an "route changed" event.

benpicco replied:
What do you mean by "route changed" event? Is there already a mechanism for this in place?

miri64 replied:
Didn't you provide a mechanism with that with the `msgbus`? Alternatively, there is also [`gnrc_netif_ipv6_t::route_info_cb`](https://doc.riot-os.org/structgnrc__netif__ipv6__t.html#aeb73dca70bd6714fd4b6592697a32ac0) specifically for routing protocols which could be extended for prefix changes.

benpicco replied:
msgbus would require a dedicated thread for this - and it looks like `route_info_cb` is called for each frame - that doesn't seem right.

miri64 replied:
> msgbus would require a dedicated thread for this

What speaks against that? Would increase the modularity of your rudimentary routing protocol after all.

miri64 replied:
> and it looks like `route_info_cb` is called for each frame - that doesn't seem right.

Only for route requests. Neighbor state changes e.g. are only called, well when the neighbor state changes. So you would just create a new hook using that function.

miri64 commented:
I'd like some more input from other contributors before merging this.

MrKevinWeiss commented:
ping @cgundogan @PeterKietzmann

MrKevinWeiss commented:
Though this is a new feature and I wouldn't mind waiting until the next release.

benpicco commented:
I think the consensus was to move this somewhere else (`sys/net/gnrc/routing`?), but I don't know how to best hook into router advertisements.

miri64 commented:
> I think the consensus was to move this somewhere else (`sys/net/gnrc/routing`?), but I don't know how to best hook into router advertisements from outside `nib.c`.

Do you have to? As far as I understand you are only interested in the upstream prefix from the PIO, not the whole RA. So can't there just be a mechanism (`msgbus`, `route_info_cb`, other callback, ...) that hands the prefix information to the new "routing" module?

miri64 reviewed (changes requested):
Please find a way to move this to `sys/net/gnrc/routing`. This is still a somewhat primitive routing protocol that you invented and not an intrinsic sub-component of IPv6 like the current module structure implies.

benpicco commented:
Will do, that's why there it still has the WIP tag :wink: 
I just wanted to clean this branch up so I can assemble a new integration branch with all the pending PRs that are needed to get this working.

benpicco commented on sys/net/gnrc/network_layer/ipv6/nib/nib.c:
```diff
@@ -86,6 +86,14 @@ static void _handle_rdnss_timeout(sock_udp_ep_t *dns_server);
 #endif
 /** @} */
 
+/* Callback to hook into Prefix Information Option */
+__attribute__ ((weak))
+void gnrc_nib_rtr_adv_pio_cb(gnrc_netif_t *netif, const ndp_opt_pi_t *pio)
+{
+    (void) netif;
+    (void) pio;
+}
```
How about this?
I moved the subnet logic to `gnrc/routing/simple_subnets/simple_subnets.c`

miri64 replied:
Seems like a good compromise. This could be extended for other options, other subscribers might be interested in. One concern I have: how would multiple subscribers be implemented?

benpicco replied:
> how would multiple subscribers be implemented?

I would deffer this issue until it actually comes up.
This is not a public API but internal to GNRC, let's not overcomplicate things for some hypothetical use case :wink: 

XFA might help us here if we don't want to waste precious RAM on a dynamic list.

miri64 commented on sys/net/gnrc/network_layer/ipv6/nib/nib.c:
```diff
@@ -86,6 +86,14 @@ static void _handle_rdnss_timeout(sock_udp_ep_t *dns_server);
 
+/* Callback to hook into Prefix Information Option */
+__attribute__ ((weak))
+void gnrc_nib_rtr_adv_pio_cb(gnrc_netif_t *netif, const ndp_opt_pi_t *pio)
```
```suggestion
void gnrc_ipv6_nib_rtr_adv_pio_cb(gnrc_netif_t *netif, const ndp_opt_pi_t *pio)
```

miri64 commented on sys/include/net/gnrc/ipv6/nib.h:
```diff
@@ -391,12 +391,25 @@ void gnrc_ipv6_nib_handle_timer_event(void *ctx, uint16_t type);
  */
 void gnrc_ipv6_nib_change_rtr_adv_iface(gnrc_netif_t *netif, bool enable);
+
+/**
+ * @brief   Immediately send a 'last router advertisement', that is a RA that
+ *          does not advertise this host as a default router.
+ *
+ *          If the @ref CONFIG_GNRC_IPV6_NIB_ADD_RIO_IN_LAST_RA option is set,
+ *          this will include a Route Information Option and nothing else.
+ *
+ * @param[in] netif     The interface on which the router advertisement should
+ *                      be sent.
+ */
+void gnrc_ipv6_nib_send_final_rtr_adv(gnrc_netif_t *netif);
```
Why is this function needed? Shouldn't it be enough to disable `RTR_ADV` on the interface?

miri64 replied:
And if not: one should rather use the `gnrc_ndp` function for that. The module for sending NDP messages is `gnrc_ndp`, `gnrc_ipv6_nib` is just a handler for these message types (and thus also sends them).

benpicco replied:
Unfortunately not. `gnrc_ipv6_nib_change_rtr_adv_iface()` will just schedule the sending of the RA, if we did just send one it can take a bit.

That means if the upstream router was restarted for some reason it can take a while for routing to the downstream network to work again.

miri64 replied:
In that case https://github.com/RIOT-OS/RIOT/pull/16536#discussion_r682475480 still applies.

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,115 @@
+        gnrc_ipv6_nib_change_rtr_adv_iface(downstream, true);
+    }
+
+    /* Disable router advertisements on upstream interface.
+     * 1. Does not confuse the upstream router to add the border router to its
+     *    default router list and
+     * 2. Includes Route Information Option in last router advertisement to
+     *    inform upstream router about downstream subnets.
+     */
+    gnrc_ipv6_nib_send_final_rtr_adv(upstream);
```
```suggestion
    /* Immediately send an RA with RIO
     */
    gnrc_pktsnip_t *ext_opts = NULL;

    while (/* use gnrc_ipv6_nib_pl_iter to get PL entries */) {
            gnrc_pktsnip_t *snip  = gnrc_ndp_opt_ri_build(&ple.pfx,
                                                          ple.pfx_len,
                                                          valid_ltime,
                                                          NDP_OPT_RI_FLAGS_PRF_NONE,
                                                          ext_opts);
            if (snip != NULL) {
                ext_opts = snip;
            } else {
                break;
            }
    }

    gnrc_ndp_rtr_adv_send(netif, NULL, &ipv6_addr_all_nodes_link_local,
                          true, ext_opts);
```

benpicco replied:
even saves us iterating over all PL entries twice

miri64 replied:
See https://github.com/RIOT-OS/RIOT/pull/16706 for how to get the `valid_ltime` ;-).

miri64 replied:
> See #16706 for how to get the `valid_ltime` ;-).

Ah, just saw that you do not even need it. Well, I think #16706 still has value for the reasons pointed out there.

benpicco replied:
Yes lifetime output was something I was sorely missing from the `nib` commands, now we can have that too!

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,119 @@
+        return;
+    }
+
+    new_prefix_len = prefix_len + 32 - __builtin_clz(subnets);
```
Some comment what you are doing here would be nice. My head spins at least.

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,119 @@
+            continue;
+        }
+
+        /* create subnet by adding interface index */
+        new_prefix.u64[0].u64 = byteorder_ntohll(prefix->u64[0]);
+        new_prefix.u64[0].u64 |= (uint64_t)subnets-- << (63 - prefix_len);
```
??! Comment says something about interface index, but all I can see is the number of interfaces (which is called `subnets` here for some reason) and the prefix length.

miri64 replied:
Maybe having a function to calculate said subnet index might make things clearer, otherwise some better explanation in the comment would be appreciated.

benpicco replied:
I mean they are just enumerated. If there are two subnets the first one will be #2 and the second one will be #1 (so just counting down the subnet number - the exact numbers are not important as long as the bit patterns are different).

miri64 replied:
Then what is the bit-shifting about?

benpicco replied:
Assume we start with `20010DB800000010/60`. For simplicity's sake let's ignore the first 56 bit, those remain constant anyway, and only focus on the last 8 bits, which are 0x10 or 0001 0000.

If we want so split this into two subnets we could have 0001 1000 and 0001 0100 - or 0x18 and 0x14.
So we just count down 2 -> 10, shift it left to the end up the prefix -> 0001 1000
1 -> 01, shift it left to the end up the prefix -> 0001 0100

miri64 replied:
Can we make this code look a bit less like black magic? In the background, `memcpy()` will most likely be used for the `uint64_t` assignment anyways, so how about we put this in a nice loop that uses the prefix length / 8 as a stop condition?

benpicco replied:
I can do without the endianness conversion, but there is still some sorcery involved: eea3f6adb987fbb717a9ad795c6fe57e10902e27

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,119 @@
+        /* create subnet by adding interface index */
+        new_prefix.u64[0].u64 = byteorder_ntohll(prefix->u64[0]);
+        new_prefix.u64[0].u64 |= (uint64_t)subnets-- << (63 - prefix_len);
+        new_prefix.u64[0] = byteorder_htonll(new_prefix.u64[0].u64);
```
Doesn't this immediately overwrite the value from the two lines before?

benpicco replied:
Yea I first convert the prefix to host byte order so it's easier to work with.
Shifting the intext onto the network byte order address would cause trouble if it ends up between two bytes.

miri64 replied:
Maybe add a comment to elevate code understanding.

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,119 @@
+        /* add route information option with new subnet */
+        ext_opts = gnrc_ndp_opt_ri_build(&new_prefix, new_prefix_len, valid_ltime,
+                                         NDP_OPT_RI_FLAGS_PRF_NONE, ext_opts);
+        if (ext_opts == NULL) {
+            DEBUG("simple_subnets: No space left in packet buffer. Not adding RIO\n");
```
Does it make sense to send the RA in that case at all?

benpicco replied:
I thought better routing to some subnets than no rounting at all :wink:

miri64 replied:
But maybe a check for `ext_opts == NULL` after the loop might make sense.

benpicco commented on doc/doxygen/src/simple_subnets.puml:
```diff
@@ -0,0 +1,40 @@
+@startuml
```
added a comment to the file

benpicco commented on doc/doxygen/src/simple_subnets.puml:
```diff
@@ -0,0 +1,41 @@
+	network level2 {
+		address = "2001:db8:0:18::/61";
+
+		router_a [address = "2001:db8:0:1e:3c27:6dff:fe25:e95d"];
```
```suggestion
		router_a [address = "2001:db8:0:18:3c27:6dff:fe25:e95d"];
```

benpicco commented:
To keep the example simple I removed the use of DHCPv6 to get a sufficiently sized prefix and instead just announce it via radvd.
I also cleaned up the `setup_taps.sh` script.

Now whether this is best placed in `examples/` I don't know - ideally it would be an automated test that verifies routing to the last node in the chain. Not sure if we have tooling to run multiple RIOT instances as part of a test though.

miri64 commented:
> […] ideally it would be an automated test that verifies routing to the last node in the chain. Not sure if we have tooling to run multiple RIOT instances as part of a test though.

In the release tests we do :-).

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	echo "creating tap interfaces"
+	i=0
+	while read -r level num; do
+		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -c %s\n" $RIOTTOOLS $i $level $num) > /dev/null;
```
I could be mistaken, but I think your script (and the overall network logic) could be greatly simplified by using `veth` for the downstream connections instead and only using the bridge for the upstream most node (I don't even think you need a bridge for that tbh). See https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth for an overview over the various virtual network device types in Linux.

benpicco replied:
will this still work with `netdev_tap`?

miri64 replied:
If the veth'd interfaces are tap interfaces, it should.

miri64 replied:
Basically, at the moment you put a switch between every one of your nodes. With veth you connect them only with a cable.

benpicco replied:
Well actually a switch is what I want - there is a routing node on each level, but it can also have multiple leaf nodes (which are the whole point of this exercise after all :wink:)

Now if I can use `veth` like a hub / bus and connect multiple nodes, that would be even better, but it looks like they always come in pairs.

miri64 commented on sys/net/gnrc/routing/simple_subnets/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = gnrc_ipv6_simple_subnets
```
Module location should match the module name, so

` sys/net/gnrc/routing/simple_subnets/` ⇒ ` sys/net/gnrc/routing/ipv6_simple_subnets/`?

miri64 commented on sys/net/gnrc/routing/simple_subnets/simple_subnets.c:
```diff
@@ -0,0 +1,156 @@
+/*
```
And

`sys/net/gnrc/routing/simple_subnets/simple_subnets.c` ⇒ `sys/net/gnrc/routing/ipv6_simple_subnets/gnrc_ipv6_simple_subnets.c`

miri64 reviewed (commented):
With the current module name I think it will cause a lot of confusion where to find it (*edit:* even though I agree that now it is more accurate). Also it's quite long. I suggest to come up with a funny name or TLA, to avoid confusion. Regardless, file naming should follow the conventions.

miri64 commented:
> [ <img alt="@benpicco" width="20" height="20" src="https://avatars.githubusercontent.com/u/1301112?s=60&amp;v=4"> ](/benpicco) ` [fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup!](/RIOT-OS/RIOT/pull/16536/commits/8932576b927ab468f0b560fda166977fc93c1bbd)`

```sh
git commit -C HEAD
```

;-).

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	echo "creating tap interfaces"
+	i=0
+	while read -r level num; do
+		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -c %s\n" $RIOTTOOLS $i $level $num) > /dev/null;
```
```suggestion
		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -c %s\n" "$RIOTTOOLS" "$i" "$level" "$num") > /dev/null;
```

miri64 replied:
```suggestion
           # shellcheck disable=SC2091 # we actually want to execute the output here.
           $(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -c %s\n" $RIOTTOOLS $i $level $num) > /dev/null;
```

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
```
Throughout this file: Please use 4 spaces instead of tab.

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+setup() {
+	# add address to loopback so we can test reaching an
+	# address outside the test net
+	sudo ip a a fd00:dead:beef::1/128 dev lo
```
(and actually unify your indentations ...)

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	while read -r level num; do
+		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -c %s\n" $RIOTTOOLS $i $level $num) > /dev/null;
+		i=$((i+1))
+	done < $1
```
```suggestion
	done < "$1"
```

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	echo "deleting tap interfaces"
+	i=0
+	while read -r level num; do
+		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -d\n" $RIOTTOOLS $i $level) > /dev/null;
```
```suggestion
		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -d\n" "$RIOTTOOLS" "$i" "$level") > /dev/null;
```

miri64 replied:
```suggestion
           # shellcheck disable=SC2091 # we actually want to execute the output here.
		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -d\n" $RIOTTOOLS $i $level) > /dev/null;
```

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	while read -r level num; do
+		$(printf -- "sudo %s/tapsetup/tapsetup -b br%s -t tap_%s -d\n" $RIOTTOOLS $i $level) > /dev/null;
+		i=$((i+1))
+	done < $1
```
```suggestion
	done < "$1"
```

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	-c)
+		;;
+	-d)
+		teardown $TOPOLOGY
```
```suggestion
		teardown "$TOPOLOGY"
```

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,64 @@
+	esac
+fi
+
+setup $TOPOLOGY
```
```suggestion
setup "$TOPOLOGY"
```

miri64 reviewed (changes requested):
Some of the `shellcheck` warnings actually do make sense. The rest can be suppressed (I think...)

miri64 commented on examples/gnrc_networking-subnets/tapology.txt:
```diff
@@ -0,0 +1,5 @@
+a	1
```
At first I wanted to point out a typo... but I see what you did there ;-)

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ *
+ * ![Example Topology](gnrc_ipv6_auto_subnets_simple.svg)
+ *
+ * The downstream network(s) get the reduced prefix via Router Advertisements and
```
This is the first time the text mentions a "reduced prefix". This should be explained.

benpicco replied:
I'll stick to 'sub-prefix' then.

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * ![Example Topology](gnrc_ipv6_auto_subnets_simple.svg)
+ *
+ * The downstream network(s) get the reduced prefix via Router Advertisements and
+ * the process repeats until the bits of prefix are exhausted. (The smallest subnet
+ * must still have a /64 prefix.)
```
```suggestion
 * the process repeats until the bits of prefix are exhausted. The smallest subnet
 * must still have a /64 prefix.
```

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * ![Example Topology](gnrc_ipv6_auto_subnets_simple.svg)
+ *
+ * The downstream network(s) get the reduced prefix via Router Advertisements and
+ * the process repeats until the bits of prefix are exhausted. (The smallest subnet
```
```suggestion
 * the process repeats until the bits of the prefix are exhausted. (The smallest subnet
```

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * the process repeats until the bits of prefix are exhausted. (The smallest subnet
+ * must still have a /64 prefix.)
+ *
+ * The downstream router will send a router advertisement with only a
```
Which "downstream router". From what I understood so far, there are many.

benpicco replied:
With this PR there can only be a single downstream router on each level.
If you want multiple downstream routers on the same link, you'll need #16750

benpicco replied:
With PR this there can only be a single downstream router on a link.
Multiple downstream routers on the same link require #16750

miri64 replied:
Then please specify, that you mean the downstream router on the same link. Downstream router can also mean the children of that router.

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * The downstream router will send a router advertisement with only a
+ * Route Information Option included to the upstream network.
+ * The Route Information Option contains the prefix of the downstream network so that
+ * upstream hosts will no longer consider hosts in this subnet on-link but instead
```
"upstream hosts"... Why are hosts (traditionally used to describe non-routing nodes) suddenly involved with routing. Given that they are upstream, shouldn't they be routers?

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ *
+ * The downstream router will send a router advertisement with only a
+ * Route Information Option included to the upstream network.
+ * The Route Information Option contains the prefix of the downstream network so that
+ * upstream hosts will no longer consider hosts in this subnet on-link but instead
+ * will use the downstream router to route to the new subnet.
```
Why is this sentence important? Up until now, the text describes the routing protocol. This sentence suddenly makes a hard break, and throws around the stirring wheel into NDP. This should be carried over better.

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * Usage
+ * =====
+ *
+ * Simply add the `gnrc_ipv6_auto_subnets` module to the nodes that should act as routers
```
How does one add a module to a node? Shouldn't it rather be added to the node's code ;-).

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+    uint8_t rem   = bits % 8;
+    int8_t shift  = 8 - rem - idx_bits;
+
+    /* first copy old prefix */
+    memcpy(out, prefix, bytes);
+
+    /* make sure unused bits are cleared */
+    if (rem) {
+        uint8_t mask = 0xff << (8 - rem);
+        out->u8[bytes] = prefix->u8[bytes] & mask;
+    } else {
+        out->u8[bytes] = 0;
+    }
```
Unless you are under timing constraints, I think this can be solved easier to understand (and I think also with smaller code footprint, as the functions is already used elsewhere in GNRC), with

```suggestion
    memset(out, 0, sizeof(*out));
    ipv6_addr_init_prefix(out, prefix, bits);
```

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+
+    DEBUG("auto_subnets: create %u subnets\n", subnets);
+
+    /* Calculate remaining prefix length. For n subnets we consume ⌊log₂ n⌋ + 1 bits.
+     * To calculate ⌊log₂ n⌋ quickly, find the position of the most significant set bit
```
Not sure if it is wise to use unicode here (especially since it can easily expressed without:

```suggestion
    /* Calculate remaining prefix length. For n subnets we consume floor(log_2 n) + 1 bits.
     * To calculate floor(log_2 n) quickly, find the position of the most significant set bit
```

miri64 replied:
Besides having Unicode here in the first place: At first sight I also confused the flooring-brackets `⌊⌋` with normal brackets `[]` because I have a very small font-size config. Then I thought something is wrong with the display and only then I realized you where using Unicode here :stuck_out_tongue_winking_eye:. I think among programers `floor()` in general is also better understood than `⌊⌋`

miri64 commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+        tmp = gnrc_ndp_opt_ri_build(&new_prefix, new_prefix_len, valid_ltime,
+                                    NDP_OPT_RI_FLAGS_PRF_NONE, ext_opts);
+        if (tmp == NULL) {
+            DEBUG("auto_subnets: No space left in packet buffer. Not adding RIO\n");
```
I might be mistaken, but I think, in case `ext_opts` was already set (as this is a loop this well may be and also with `_remove_old_prefix()`), you need to release `ext_opts` in this case. Otherwise there is a potential packet buffer leak here.

benpicco replied:
That's why I used the `tmp` variable. `ext_opts` is only overwritten if `tmp` is not NULL, then `gnrc_ndp_rtr_adv_send()` is going to take care of it.

miri64 replied:
Yeah, but in previous iterations or in `_remove_old_prefix()` `ext_opts` might have been allocated which is free'd nowhere now, right?

miri64 replied:
Ah,... you are not bailing here! But maybe add a break instead of trying to fill up an already filled packet buffer?

miri64 replied:
Ping?

benpicco replied:
I thought it would be still worthwhile to configure the downstream networks even if the packet buffer is currently exhausted to inform upstream.

We might also bail out here to make the breakage more visible.

miri64 replied:
> I thought it would be still worthwhile to configure the downstream networks even if the packet buffer is currently exhausted to inform upstream.
> 
> We might also bail out here to make the breakage more visible.

Might make sense. And there is a chance, that some other module gives up resources again, so maybe we do not need to `break;` here. Ok!

miri64 commented:
> ![](https://raw.githubusercontent.com/RIOT-OS/RIOT/98db19c53d5c04937b4ae03ec61c749805f474ab/doc/doxygen/src/gnrc_ipv6_auto_subnets_simple.svg)

Can either the edges or the IIDs be a different color? I have a hard time recognizing the IID values on the edges.

miri64 commented:
> Can either the edges or the IIDs be a different color? I have a hard time recognizing the IID values on the edges.

Alternatively, give the addresses a white background. I think that makes them more readable in general (including the address @ uplink)

benpicco commented:
Sure, I made them gray to indicate that they are not important for what I want to show (some are just the result of keyboard mashing anyway), but I can add a background color.

miri64 commented:
> Sure, I made them gray to indicate that they are not important for what I want to show (some are just the result of keyboard mashing anyway), but I can add a background color.

Sometimes trying to hide something makes it more obvious ;-)

![](https://media.giphy.com/media/BK8hW6YBpnvSU/giphy.gif?cid=ecf05e479nhs8pjq6uce5f77vqh5zmrdijuqm987v18qvct1&rid=giphy.gif&ct=g)

miri64 commented:
Wait... uplink has `2001:db8::/64`? Then `2001:db8:0:8::/61` is not part of that subnet and packets toward it won't be forwarded by the uplink router (best case) and you are spoofing foreign IP addresses when sending into the wider internet (worst case).

benpicco commented:
The uplink is 2001:dbc::/64

miri64 commented:
> The uplink is 2001:db**c**::/64

Ok, so not even the same subnet as you use from the gateway oO? Shouldn't it rather be `2001:db8::/59` or something (I get /56 from my ISP)?

miri64 commented:
> (Ok, that's outside the range allocated for documentation - maybe better make that `2001:db8:1::/64` then. Or leave it out entirely doesn't really matter where the prefix came from)

Ahhh (post-edit) the "uplink" is supposed to be another downlink from Gateway? Then maybe leave it out altogether. Seems to cause more confusion than having any purpose for understanding the protocol.

miri64 commented:
> Can either the edges or the IIDs be a different color? I have a hard time recognizing the IID values on the edges.

Maybe also use he same style sheet used in [other SVGs in the doc](https://github.com/RIOT-OS/RIOT/blame/9294a31856c328efb9a25ae28e607b54416c210c/doc/doxygen/src/riot-structure.svg#L381-L467) (if possible)?

miri64 commented:
(or at least the same color scheme)

benpicco commented on sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c:
```diff
@@ -0,0 +1,221 @@
+ * routing node with this module will automatically configure subnets from it for
+ * each downstream interface.
+ *
+ * There can only be a single routing node on each level of the network but an
+ * arbitrary number of leaf nodes.
```
Yes that's the idea

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -7,58 +7,59 @@ RIOTTOOLS=$RIOTBASE/dist/tools
-	done < "$1"
+    echo "deleting tap interfaces"
+    i=0
+    while read -r level num; do
```
Also needs a filter for the comments now, I believe.

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,65 @@
+    echo "creating tap interfaces"
+    i=0
+    sed '/^#/d' "$1" | while read -r level num; do
+        # shellcheck disable=SC2091 # we actually want to execute the output here.
```
Mhhh apparently the comment makes a problem. Can you add it above the shellcheck directive?

benpicco commented:
Unfortunately nwdiag does [not support styling](https://forum.plantuml.net/14479/style-allow-style-on-nwdiag-diagram) yet. 

I can do my own styling with `sed`
```bash
#!/bin/sh
BGCOLOR=3fa687
FGCOLOR=275a4b

sed -i s/A80036/$BGCOLOR/g $1
sed -i s/FEFECE/$FGCOLOR/g $1
sed -i s/000000/$FGCOLOR/g $1
sed -i s/Serif/Miso/g $1
sed -i -E s/font-size=\"11\"/font-size=\"16\"/g $1
```

but the result looks a bit abominable if the *Miso* font is not installed

![](https://gist.githubusercontent.com/benpicco/ae0510998397789056857d7d9736b003/raw/efd59bf1ae20612690ef2b45d3385cd15b795bf8/gnrc_ipv6_auto_subnets_simple.svg)

With the font it's a bit better:

![gnrc_ipv6_auto_subnets_simple](https://user-images.githubusercontent.com/1301112/131524476-33561c25-2345-4a6a-b069-6a6776d4229c.png)

miri64 commented:
> but the result looks a bit abominable if the _Miso_ font is not installed

Yeah, I was expecting that. The other SVGs actually include it in their CSS. But does it work without the Miso font?

miri64 commented:
But let's not make things more complicated than they need to. If styling is not supported by the tool, then styling is not supported. No need to add yet another script to fix that.

miri64 commented:
Please squash while I try to figure out how to test this in `native` ;-)

miri64 commented on examples/gnrc_networking-subnets/Makefile:
```diff
@@ -0,0 +1,49 @@
+# name of your application
+APPLICATION = gnrc_networking-subnets
```
Here and directory name: can we stick to underscore's please?

miri64 commented on examples/gnrc_networking-subnets/main.c:
```diff
@@ -0,0 +1 @@
+../gnrc_networking/main.c
```
File is missing `x` permissions.

benpicco replied:
I think you meant the .sh file

miri64 replied:
Yepp. Sorry. Too many files ^^

miri64 commented on examples/gnrc_networking-subnets/setup_taps.sh:
```diff
@@ -0,0 +1,66 @@
+
+RIOTBASE=../..
+RIOTTOOLS=$RIOTBASE/dist/tools
+TOPOLOGY=tapology.txt
```
```console
./examples/gnrc_networking-subnets/setup_taps.sh 
no such file: tapology.txt
```

:-/

benpicco replied:
uh did I forget to commit this

benpicco replied:
should be fixed now.

miri64 reviewed (commented):
Examples must have a README.md!

miri64 commented:
Also 

```console
$ PORT=tap_a0 make term
/home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking-subnets/bin/native/gnrc_networking-subnets.elf tap_a0
usage: /home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking-subnets/bin/native/gnrc_networking-subnets.elf <tap interface 1> <tap interface 2> [-i <id>] [-d] [-e|-E] [-o] [-c <tty>] [-g <gpiochip>] [-i <id>] [-d] [-e|-E] [-o] [-c <tty>] [--eui64 <eui64> …]

help: /home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking-subnets/bin/native/gnrc_networking-subnets.elf -h
[…]
```

...

benpicco commented:
Ok, let me pen up a README.md

benpicco commented:
> Also
> 
> ```
> $ PORT=tap_a0 make term
> ```

The routers all have two interfaces, so 

    PORT="tap_a0 tap_b1" make term
    PORT="tap_b0 tap_c1" make term
    …

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -0,0 +1,78 @@
+a single router on each link.
+Routers can have multiple interfaces to connect different downlinks.
+
+![](https://raw.githubusercontent.com/RIOT-OS/RIOT/master/doc/doxygen/src/gnrc_ipv6_auto_subnets_simple.svg)
```
IIRC 
```suggestion
![](/RIOT-OS/RIOT/master/doc/doxygen/src/gnrc_ipv6_auto_subnets_simple.svg)
```

should work as well.

benpicco replied:
Not if the file is viewed locally

miri64 replied:
Ah yes. But 
```suggestion
![](../../doc/doxygen/src/gnrc_ipv6_auto_subnets_simple.svg)
```

should also work :-)

benpicco replied:
I thought so too, but at least [grip](https://github.com/joeyespo/grip) does not like it.

miri64 replied:
Then why do we need to honor this?

benpicco replied:
Yea you're right

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -0,0 +1,78 @@
+
+### Obtaining the prefix
+
+#### via router advertisements
```
```suggestion
#### Via router advertisements
```

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -0,0 +1,78 @@
+If you can configure your router to advertise a larger prefix instead, this
+will be the easiest solution.
+
+#### via DHCPv6
```
```suggestion
#### Via DHCPv6
```

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -0,0 +1,78 @@
+If you can't change the configuration of your router, but your router supports
+DHCPv6 IA_PD, you can use this to request a larger prefix.
+
+This requires a gateway node that does uses the `gnrc_dhcpv6_client_simple_pd`
```
```suggestion
This requires a gateway node that uses the `gnrc_dhcpv6_client_simple_pd`
```

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -0,0 +1,78 @@
+
+It is also possible to connect non-routing leaf nodes with a single interface:
+
+    make term LEAF=1 PORT=tap_b2
```
I get

```
Terminal program /home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking_subnets/bin/native/gnrc_networking-subnets_leaf.elf is required but not found in PATH.  Aborting.
```

with that.

benpicco replied:
Ok, will replace that with

    make all term LEAF=1 PORT=tap_b2

miri64 reviewed (commented):
Application works as described in the README

![image](https://user-images.githubusercontent.com/675644/131842203-6c637a7e-fe99-41f5-93a1-f9725996e9ba.png)

However, there is a problem with the `LEAF` corner case

miri64 commented on examples/gnrc_networking_subnets/setup_taps.sh:
```diff
@@ -0,0 +1,68 @@
+        i=$((i+1))
+    done
+
+    # start radvd with a large prefix
+    sudo "$RIOTTOOLS"/radvd/radvd.sh -c br0 $PREFIX
```
Would be nice if `PREFIX::1` could be added to `br0`, so one can test this by pinging the host system.
```suggestion
    sudo ip addr add ${PREFIX/::\//::1\/} dev br0
    # start radvd with a large prefix
    sudo "$RIOTTOOLS"/radvd/radvd.sh -c br0 $PREFIX
```

benpicco replied:
The setup script will already add `fd00:dead:beef::1` to `lo` for that

miri64 replied:
> The setup script will already add `fd00:dead:beef::1` to `lo` for that

But why not go with common convention and have the upstream gateway be `PREFIX::1`?

miri64 commented on examples/gnrc_networking_subnets/README.md:
```diff
@@ -39,7 +39,7 @@ Start more `native` instances to simulate a cascading tree network:
 It is also possible to connect non-routing leaf nodes with a single interface:
 
-    make term LEAF=1 PORT=tap_b2
+    make all term LEAF=1 PORT=tap_b2
```
Better go with 
```suggestion
    make flash term LEAF=1 PORT=tap_b2
```

this way `term` is executed after the build, even with parallel make ;-).

benpicco replied:
`make flash` on `native` feels weird

miri64 commented:
All steps in the README now work as advertised.

miri64 reviewed (commented):
Please squash!

miri64 reviewed (approved):
ACK!
